### PR TITLE
[AKS] Refactor acs module: refactor aks sub-command create part 1

### PR DESCRIPTION
**Pull request by FumingZhang**
**Description**

[DO NOT MERGE] Refactor aks sub-command create.
* Please refer to the sample PR#3743 in aks-preview to verify the effect of reuse.
* Please refer to this wiki for the details of the modification.

**Testing Guide**

**History Notes**

[Component Name 1] BREAKING CHANGE: az command a: Make some customer-facing breaking change.
[Component Name 2] az command b: Add some customer-facing feature.

---

This checklist is used to make sure that common guidelines for a pull request are followed.

- [ ] The PR title and description has followed the guideline in [Submitting Pull Requests](https://github.com/Azure/azure-cli/tree/dev/doc/authoring_command_modules#submitting-pull-requests).

- [ ] I adhere to the [Command Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/command_guidelines.md).

- [ ] I adhere to the [Error Handling Guidelines](https://github.com/Azure/azure-cli/blob/dev/doc/error_handling_guidelines.md).

### CODE REVIEW

**Comment by yonzhan** [comment]
AKS

File: src/azure-cli/azure/cli/command_modules/acs/custom.py [hunk 1]
```diff
@@ -1951,6 +1956,202 @@ def _add_virtual_node_role_assignment(cmd, result, vnet_subnet_id):
...
+        pass
+
+
+class AKSCreateDecorator:
+    def __init__(self, cmd, client, models, raw_parameters, resource_type=ResourceType.MGMT_CONTAINERSERVICE):
+        self.cmd = cmd
+        self.client = client
+        self.models = models
+        self.param = AKSCreateParameters(raw_parameters)
+        self.context = AKSCreateContext()
```
Comment (andyliuliming) [inline src/azure-cli/azure/cli/command_modules/acs/custom.py:2069]: could you please add some comments describing which kind of data should be stored to the context.
so people can follow.

File: src/azure-cli/azure/cli/command_modules/acs/custom.py [hunk 2]
```diff
@@ -1951,6 +1956,202 @@ def _add_virtual_node_role_assignment(cmd, result, vnet_subnet_id):
...
+        # `resource_type` is used to dynamically find the model (of a specific api version) provided by the
+        # containerservice SDK, most models have been passed through the `modles` parameter (instantiatied
+        # from `AKSCreateModels` (or `PreviewAKSCreateModels` in aks-preview), where resource_type (i.e.,
+        # api version) has been specified), a very small number of models are instantiated through internal
+        # functions, one use case is that `api_server_access_profile` is initialized by function
+        # `_populate_api_server_access_profile` defined in `_helpers.py`
+        self.resource_type = resource_type
+
+    def get_variable_from_param_and_context(self, variable_name, prefer_context=False):
```
Comment (andyliuliming) [inline src/azure-cli/azure/cli/command_modules/acs/custom.py:2078]: comments.

File: src/azure-cli/azure/cli/command_modules/acs/custom.py [hunk 3]
```diff
@@ -1951,6 +1956,202 @@ def _add_virtual_node_role_assignment(cmd, result, vnet_subnet_id):
...
+            variable = getattr(self.context, variable_name, None)
+            if variable is None:
+                variable = getattr(self.param, variable_name, None)
+        else:
+            variable = getattr(self.param, variable_name, None)
+            if variable is None:
+                variable = getattr(self.context, variable_name, None)
+        return variable
+
+    def check_params(self):
```
Comment (andyliuliming) [inline src/azure-cli/azure/cli/command_modules/acs/custom.py:2090]: comments.

**Review by andyliuliming — CHANGES_REQUESTED** [review]

File: src/azure-cli/azure/cli/command_modules/acs/custom.py [hunk 4]
```diff
@@ -1951,6 +1956,202 @@ def _add_virtual_node_role_assignment(cmd, result, vnet_subnet_id):
...
+    def __init__(self):
+        pass
+
+
+class AKSCreateDecorator:
+    def __init__(self, cmd, client, models, raw_parameters, resource_type=ResourceType.MGMT_CONTAINERSERVICE):
+        self.cmd = cmd
+        self.client = client
+        self.models = models
+        self.param = AKSCreateParameters(raw_parameters)
```
Comment (andyliuliming) [inline src/azure-cli/azure/cli/command_modules/acs/custom.py:2068]: can we try to add some linting rule that check the param won't be modified.

**Review by andyliuliming — CHANGES_REQUESTED** [review]

File: src/azure-cli/azure/cli/command_modules/acs/custom.py [hunk 5]
```diff
@@ -1951,6 +1957,326 @@ def _add_virtual_node_role_assignment(cmd, result, vnet_subnet_id):
...
+        # `True` or the cached value is the same as the original value.
+        @functools.wraps(func)
+        def wrapper(self, force_update=False, keep_cache=False, read_raw=False, default_value=None, **kwargs):
+            # function name should follow the naming method of `get_xxx`, and `xxx` is the parameter name
+            parameter_name = func.__name__[4:]
+            kwargs["parameter_name"] = parameter_name
+
+            # fetch raw and cached value
+            raw_value = self.get_param(parameter_name, read_raw=True, default_value=default_value)
+            cached_value = self.get_param(parameter_name, read_raw=False, default_value=default_value)
```
Comment (andyliuliming) [inline src/azure-cli/azure/cli/command_modules/acs/custom.py:2118]: what's the benifits?

**Review by andyliuliming — CHANGES_REQUESTED** [review]

**Comment by zhoxing-ms** [comment]
What is the difference between PR #19259 and this PR? If they are the same PR, could you close it?

**Comment by FumingZhang** [comment]
The modification has been replaced by PR #19259, closed,

**Pull request closed without merging**